va/http.go

```diff
@@ -0,0 +1,424 @@
+		deadline = deadline.Add(-10 * time.Millisecond)
+	}
+	ctx, cancel := context.WithDeadline(ctx, deadline)
+	defer cancel()
+
+	// Create a target for the host, port and path
+	target, err := va.newHTTPValidationTarget(ctx, host, va.httpPort, path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Set up the initial validation request and a base validation record
+	initialReq, baseRecord, err := va.setupHTTPValidation(ctx, nil, target)
+	if err != nil {
+		return nil, []core.ValidationRecord{}, err
+	}
+
+	va.log.AuditInfof("Attempting to validate HTTP-01 for %q with GET to %q",
+		initialReq.Host, initialReq.URL.String())
+
+	// Create a closure around records & numRedirects we can use with a HTTP
+	// client to process redirects per our own policy (e.g. resolving IP
+	// addresses explicitly, not following redirects to ports != [80,443], etc)
+	records := []core.ValidationRecord{baseRecord}
+	numRedirects := 0
+	processRedirect := func(req *http.Request, via []*http.Request) error {
+		va.log.Infof("processing a HTTP redirect from the server to %q\n", req.URL.String())
+		// Only process up to maxRedirect redirects
+		if numRedirects > maxRedirect {
+			return berrors.ConnectionFailureError("Too many redirects")
+		}
+		numRedirects++
+		va.metrics.http01Redirects.Inc()
+
+		// Extract the redirect target's host and port. This will return an error if
+		// the redirect request scheme, host or port is not acceptable.
+		redirHost, redirPort, err := va.extractRequestTarget(req)
+		if err != nil {
+			return err
+		}
+
+		redirPath := req.URL.Path
+		// Setup a validation target for the redirect host. This will resolve IP
+		// addresses for the host explicitly.
+		redirTarget, err := va.newHTTPValidationTarget(ctx, redirHost, redirPort, redirPath)
+		if err != nil {
+			return err
+		}
+
+		// Mutate the existing redirect request to use a URL and Host based on the
+		// explicitly resolved target IPs. This will also give us a validationRecord
+		// for the redirect which we should append to the records.
+		_, redirRecord, err := va.setupHTTPValidation(ctx, req, redirTarget)
+		records = append(records, redirRecord)
+		if err != nil {
+			return err
+		}
+		va.log.Infof("following redirect to host %q url %q\n", req.Host, req.URL.String())
+		return nil
+	}
+
+	// Create a new HTTP client and check HTTP redirects it encounters with
+	// processRedirect
+	client := newHTTPClient()
+	client.CheckRedirect = processRedirect
+
+	// Make the initial validation request. This may result in redirects being
+	// followed.
+	httpResponse, err := client.Do(initialReq)
+	// If there was an error we may need to try a fallback.
+	if err != nil {
+		// Calling To4 on an IPv6 address returns nil and can be used to
+		// differentiate IPv6/IPv4 net.IP instances
+		triedIPv6 := baseRecord.AddressUsed.To4() == nil
+		if !triedIPv6 {
+			// If the base validation record indicates we used an IPv4 address then we
```

Comment:
I think by the time we get here, `newHTTPValidationTarget` has already ensured that `target.next` has at most two entries, with the first one potentially being IPv6. So these checks can probably be removed in favor of just advancing to the next item in `next`.

Speaking of which, I commented above on `.IP()`, but it's a little confusing that advancing to the next item in `next` happens as a side effect of calling `.IP()`, inside `setupHTTPValidation`. I don't have a strong suggestion for how to untangle that bit of state, but I wanted to mention it in case you have any ideas.

Reply:
> I think by the time we get here, newHTTPValidationTarget has already ensured that target.next has at most two entries, with the first one potentially being IPv6. So these checks can probably be removed in favor of just advancing to the next item in next.

Good point!

> Speaking of which, I commented above on .IP(), but it's a little confusing that advancing to the next item in next happens as a side effect of calling .IP(), inside setupHTTPValidation. I don't have a strong suggestion for how to untangle that bit of state, but I wanted to mention it in case you have any ideas.

I agree its not super clear how those two things interact. I took a crack at untangling it slightly in 4c8f482 by separating the two things `httpValidationTarget.ip()` was doing previously: advancing through next (while updating tried) and returning the current IP to use. I think it makes it clearer if there's an explicit call to `nextIP()` separate ahead of calling `setupHTTPValidation()` and its call to `ip()`.

Does 4c8f482 look better to you?